server/src/migrations/20241217092523-migration-applications-to-applicants.ts

```diff
@@ -0,0 +1,59 @@
+import { ObjectId } from "bson"
+import { IApplicant, ZApplicant } from "shared"
+
+import { getDbCollection } from "@/common/utils/mongodbUtils"
+
+import { asyncForEach } from "../common/utils/asyncUtils"
+
+type ApplicationAggregate = {
+  _id: string
+  firstname: string
+  lastname: string
+  phone: string
+  email: string
+  ids: string[]
+  last_connection: Date
+}
+
+export const up = async () => {
+  const applications = (await getDbCollection("applications")
+    .aggregate([
+      {
+        $group: {
+          _id: "$applicant_email",
+          firstname: { $first: "$applicant_first_name" },
+          lastname: { $first: "$applicant_last_name" },
+          phone: { $first: "$applicant_phone" },
+          email: { $first: "$applicant_email" },
+          ids: { $push: "$_id" },
+          last_connection: { $last: "$created_at" },
+        },
+      },
+    ])
+    .toArray()) as ApplicationAggregate[]
+  const stat = { error: 0, success: 0, total: applications.length }
+
+  await asyncForEach(applications, async (application) => {
+    const { firstname, lastname, email, phone, last_connection, ids } = application
+    const now = new Date()
+    const applicant: IApplicant = {
+      _id: new ObjectId(),
+      firstname,
+      lastname,
+      phone,
```

Comment:
Ca va être trop conséquent sans streamer (plus de 1,4M d'entrées en base). Soit il faut ajouter des merge dans l'aggregate, soit streamer sans le toArray avec un 

`for await (application of applications)`

Reply:
c'est aggrégé par email, de 1,4M on passe à 142K. Je peux passer avec un for await oui

Reply:
j'ai fait la modification mais je ne pense pas que ça change grand chose. Pour avoir la complétion typescript je dois utiliser le toArray()